### DISCUSSION TITLE
How to force certain levels to be expressed in a query?

### DISCUSSION THREAD

**Discussion by amhatre** [atoti/atoti#521] [Q&A] [answered]
Hello, I would like to configure a measure to return a value only when certain levels are expressed in a query. 
For eg. if I have a dataset like below, I would like a measure `Price` which will return a value only when Fund and Security levels are expressed. 

Fund | Security | Quantity | Price
-- | -- | -- | --
INX | TSLA | 10 | 100
DJI | AAPL | 20 | 40

Could you advise on how to configure such a measure?

One way to do this in the Java api would be to have Fund and Security as leaf levels, and use `.doNotAggregateAbove()`.
Is there anything equivalent of `.doNotAggregateAbove()` in the atoti python api?

**Comment by Acriss** [comment]
Hello !
One way to do this is to use `tt.where(...)` with a condition where you check that `l["Fund"] != None`.
Since `null` cannot be a member of a hierarchy, checking that a level is not "None" is equivalent to checking that it's expressed in the query.

I am unsure if you can do this same thing with `tt.filter`, but you might try the same condition and see if it works. I'll try to check it as well on my side.

**Reply by rcolle** [reply]
`tt.where` is the way to go. `tt.filter` applies conditions but does not mask any data.

**Comment by bloussou** [comment, marked as answer]
Hello,
Aren't you looking for [`tt.value(table["Price"], levels=[l["Fund"], l["Security"]])`](https://docs.atoti.io/latest/lib/atoti.html#atoti.value) because the doc of levels is 
```
The levels that must be expressed for this measure to possibly be non-null.

        When ``None``, the measure will also be ``None`` if the levels corresponding to the keys of *column*'s table are not expressed.

        Passing an empty collection propagate the value on all levels when possible.
```

**Comment by amhatre** [comment]
Thanks guys! Both the solutions work.

@bloussou Is `tt.value` designed to be used for measures where the numerical column is in the reference store? For eg. the "Purchase price" column [here](https://docs.atoti.io/latest/lib/atoti.html#atoti.value). 
In my case the dataset I mentioned above is the base store. Regardless, it works for me.

Following Pierre's suggestion, my measure is as follows - 
```
m["Price"] = tt.where((l["Fund"] != None) & (l["Security"] != None), m["Price.SUM"], None)
```
In my case, I guess it is better to use `tt.value` as there is no conditional check, and simply a datastore lookup?

**Reply by bloussou** [reply]
You can also use it on the base store without issue, it is designed to support partial join, full join and selection fields. `tt.value()` is `.per(levels).doNotAggregateAbove()` as you are looking for so I think it can be faster than conditional checking but I'm not sure about this.

**Reply by amhatre** [reply]
Thanks @bloussou
